File: src/main/java/pt/haslab/dql/tictactoe/game/GameMain.java
package pt.haslab.dql.tictactoe.game;

import pt.haslab.dql.tictactoe.agents.*;

/**
 * Tic-Tac-Toe main class. Inspired by:
 * https://www.ntu.edu.sg/home/ehchua/programming/java/JavaGame_TicTacToe.html
 * Created by nunomachado on 05/08/17.
 */
public class GameMain {
    public Board board;            // the game board
    public GameState currentState; // the current state of the game (of enum GameState)
    public Seed currentPlayer;     // the current player (of enum Seed)

    /** Constructor to setup the game */
    public GameMain() {
        board = new Board();  // allocate game-board

        // Initialize the game-board and current status
        initGame();
    }

    /** Initialize the game-board contents and the current states */
    public void initGame() {
        board.init();  // clear the board contents
        currentPlayer = Seed.CROSS;       // CROSS plays first
        currentState = GameState.PLAYING; // ready to play
    }

    /**
     * Implements the logic of a tic-tac-toe game
     * played by two agents
     * @param aX
     * @param aO
     */
    public int playGame(Agent aX, Agent aO) {
        Agent agent;
        boolean validAction = false;
        int result = 0; // X wins = 1, O wins = -1, Draw = 0

        do {
            if(currentPlayer == Seed.CROSS){
                System.out.println("\nPlayer 'X', enter your move (row[1-3] column[1-3]): ");
                agent = aX;
            }
            else{
                System.out.println("\nPlayer 'O', enter your move (row[1-3] column[1-3]): ");
                agent = aO;
            }

            //execute move by agent
            do {
                String a = agent.act(board);
                String[] actions = a.split(" ");
                System.out.println(">> "+a);
                validAction = playerMove(currentPlayer, Integer.valueOf(actions[0]), Integer.valueOf(actions[1]));
            } while (!validAction);

            //update board and currentState
            board.paint();
            updateGame(currentPlayer);

            // Print message if game-over
            if (currentState == GameState.CROSS_WON) {
                System.out.println("'X' won! Bye!");
                result = 1;
            } else if (currentState == GameState.NOUGHT_WON) {
                System.out.println("'O' won! Bye!");
                result = -1;
            } else if (currentState == GameState.DRAW) {
                System.out.println("It's Draw! Bye!");
            }

            // Switch player
            currentPlayer = (currentPlayer == Seed.CROSS) ? Seed.NOUGHT : Seed.CROSS;
        } while (currentState == GameState.PLAYING);  // repeat until game-over

        return result;
    }

    /** The player with "theSeed" makes one move with input validation.
     Update Cell's content, Board's currentRow and currentCol. */
    public boolean playerMove(Seed theSeed, int r, int c) {
        int row = r - 1;
        int col = c - 1;
        if (row >= 0 && row < Board.ROWS && col >= 0 && col < Board.COLS
                && board.cells[row][col].content == Seed.EMPTY) {
            board.cells[row][col].content = theSeed;
            board.currentRow = row;
            board.currentCol = col;
            return true; // input okay, exit loop
        } else {
            System.out.println("This move at (" + (row + 1) + "," + (col + 1)
                    + ") is not valid. Try again...");
        }
        return false;
    }

    /**
     * Update the currentState after the player with "theSeed" has moved.
     * Return corresponding reward (1 : cross won, -1 : nought won, 0.5 : draw, 0 : otherwise)
     * */
    public double updateGame(Seed theSeed) {
        if (board.hasWon(theSeed)) {  // check for win
            currentState = (theSeed == Seed.CROSS) ? GameState.CROSS_WON : GameState.NOUGHT_WON;
            if(currentState == GameState.CROSS_WON)
                return 1;
            else
                return  -1;
        } else if (board.isDraw()) {  // check for draw
            currentState = GameState.DRAW;
            return 0;
        }
        // Otherwise, no change to current state (still GameState.PLAYING).
        return 0;
    }
}

File: tictactoe/game/game_main.py
from enum import Enum
from typing import Tuple

from tictactoe.agents.agent import Agent
from tictactoe.game.board import Board, Seed
from tictactoe.game.moves import Moves


class GameState(Enum):
    PLAYING = 1
    DRAW = 2
    CROSS_WON = 3
    NOUGHT_WON = 4


class GameMain:
    def __init__(self, agent_X: Agent, agent_O: Agent) -> None:
        """Constructor to setup the game."""
        self.board = Board()
        self.current_state = GameState.PLAYING
        self.current_player = Seed.CROSS
        self.agent_X = agent_X
        self.agent_O = agent_O

        self.board.init()

    def play_game(self) -> int:
        """Implements the logic of a tic-tac-toe game played by two agents."""
        print(f"\n===== GAME =====")
        while self.current_state == GameState.PLAYING:
            agent = self.agent_X if self.current_player == Seed.CROSS else self.agent_O
            print(
                f"\nPlayer '{self.current_player.value}' ({agent.__class__.__name__}), enter your move (row[1-3] column[1-3]): "
            )

            # get valid move from agent
            valid_move = False
            while not valid_move:
                move = agent.make_move(self.board)
                print(f">> {move}")
                try:
                    r, c = Moves.move_to_tuple(move)
                    _, reward, _ = self.apply_move(r, c, Seed.CROSS) # assume target_seed is CROSS
                    valid_move = True
                except ValueError:
                    print(f"Move {move} is not valid. Try again...")

            # update board and current state
            self.board.paint()
    
        return reward
    
    def apply_move(self, row: int, col: int, target_seed: Seed) -> Tuple[Board, float, bool]:
        """Take an move, update the game state, and return the next state, reward, and done flag."""
        valid_move = self.validate_move(self.current_player, row, col)
        
        if not valid_move:
            raise ValueError(f"Invalid move: {row} {col}")

        reward = self.update_game(self.current_player, target_seed)
        done = self.current_state != GameState.PLAYING

        # Switch player if the game is not done
        if not done:
            self.current_player = (
                Seed.NOUGHT if self.current_player == Seed.CROSS else Seed.CROSS
            )

        return self.board, reward, done

    def validate_move(self, seed: Seed, r: int, c: int) -> bool:
        """The player with 'seed' makes one move with input validation."""
        row = r - 1
        col = c - 1
        if (
            0 <= row < Board.ROWS
            and 0 <= col < Board.COLS
            and self.board.cells[row][col].content == Seed.EMPTY
        ):
            self.board.cells[row][col].content = seed
            self.board.current_row = row
            self.board.current_col = col
            return True
        else:
            return False

    def update_game(self, player_seed: Seed, target_seed: Seed ) -> float:
        """Update the currentState after the player with 'player_seed' has moved.
        Return corresponding reward according to the target_seed (1 if target_seed won, -1 if target_seed lost, 0 if draw or otherwise)
        """
        if self.board.has_won(player_seed):
            self.current_state = (
                GameState.CROSS_WON if player_seed == Seed.CROSS else GameState.NOUGHT_WON
            )
            if self.current_state == GameState.CROSS_WON:
                print("'X' won!")
                return 1 if target_seed == Seed.CROSS else -1
            else:
                print("'O' won!")
                return 1 if target_seed == Seed.NOUGHT else -1
        elif self.board.is_draw():
            self.current_state = GameState.DRAW
            print("It's a Draw!")
            return 0
        
        # Small negative reward for each move to encourage faster wins
        return -0.01